adapter: move peek optimization off coord thread

mjibson opened the pull request:
Move optimization off of the coordinator thread so that slow optimize phases don't block other messages.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

mjibson commented:
Draft because I think this needs a bit more thought around correctness. What happens if the set of indexes or objects changes after the optimization starts, and they are no longer valid? This probably needs to check the catalog transient revision and re-plan if things have changed. Due to all of the current staging code this feels...very achievable?

philip-stoev commented:
@mjibson please add me as a reviewer once this is ready. I will build a stress test with concurrent DDL

mjibson commented:
This is ready for review and testing now. It's been rebased with the change that re-plans if the catalog has changed.

philip-stoev commented:
@mjibson  I will try to work on that today, thank you

philip-stoev commented:
Item No 1. A simple concurrent workload causes a panic in this PR and not in the parent commit:

```
thread 'coordinator' panicked at 'no entry found for key', src/adapter/src/catalog.rs:505:10
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:64:14
   2: core::panicking::panic_display
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:147:5
   3: core::panicking::panic_str
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:131:5
   4: core::option::expect_failed
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/option.rs:1924:5
   5: <core::option::Option<&mz_adapter::catalog::CatalogEntry>>::expect
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/option.rs:786:21
   6: <alloc::collections::btree::map::BTreeMap<mz_repr::global_id::GlobalId, mz_adapter::catalog::CatalogEntry> as core::ops::index::Index<&mz_repr::global_id::GlobalId>>::index
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/alloc/src/collections/btree/map.rs:2216:23
   7: <mz_adapter::catalog::CatalogState>::get_entry
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/catalog.rs:505:10
   8: <mz_adapter::catalog::Catalog>::get_entry
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/catalog.rs:4398:9
   9: mz_adapter::coord::introspection::auto_run_on_introspection::<alloc::collections::btree::set::BTreeSet<mz_repr::global_id::GlobalId>>::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord/introspection.rs:61:21
  10: core::iter::traits::iterator::Iterator::all::check::<mz_repr::global_id::GlobalId, mz_adapter::coord::introspection::auto_run_on_introspection<alloc::collections::btree::set::BTreeSet<mz_repr::global_id::GlobalId>>::{closure#0}>::{closure#0}
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/iter/traits/iterator.rs:2604:20
  11: <core::iter::adapters::peekable::Peekable<alloc::collections::btree::set::IntoIter<mz_repr::global_id::GlobalId>> as core::iter::traits::iterator::Iterator>::try_fold::<(), core::iter::traits::iterator::Iterator::all::check<mz_repr::global_id::GlobalId, mz_adapter::coord::introspection::auto_run_on_introspection<alloc::collections::btree::set::BTreeSet<mz_repr::global_id::GlobalId>>::{closure#0}>::{closure#0}, core::ops::control_flow::ControlFlow<()>>
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/iter/adapters/peekable.rs:98:30
  12: <core::iter::adapters::peekable::Peekable<alloc::collections::btree::set::IntoIter<mz_repr::global_id::GlobalId>> as core::iter::traits::iterator::Iterator>::all::<mz_adapter::coord::introspection::auto_run_on_introspection<alloc::collections::btree::set::BTreeSet<mz_repr::global_id::GlobalId>>::{closure#0}>
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/iter/traits/iterator.rs:2607:9
  13: mz_adapter::coord::introspection::auto_run_on_introspection::<alloc::collections::btree::set::BTreeSet<mz_repr::global_id::GlobalId>>
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord/introspection.rs:60:41
  14: <mz_adapter::coord::Coordinator>::peek_stage_validate
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord/sequencer/inner.rs:1917:19
  15: <mz_adapter::coord::Coordinator>::sequence_peek_stage::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord/sequencer/inner.rs:1862:40
  16: <mz_adapter::coord::Coordinator>::sequence_peek::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord/sequencer/inner.rs:1846:9
  17: <mz_adapter::coord::Coordinator>::sequence_peek::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord/sequencer/inner.rs:1831:5
  18: <mz_adapter::coord::Coordinator>::sequence_plan::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord/sequencer.rs:235:66
  19: <mz_adapter::coord::Coordinator>::sequence_plan::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord/sequencer.rs:53:5
  20: <mz_adapter::coord::Coordinator>::handle_message::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord/message_handler.rs:103:21
  21: <mz_adapter::coord::Coordinator>::serve::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord.rs:1334:37
  22: <tokio::runtime::park::CachedParkThread>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>::{closure#0}
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/park.rs:283:63
  23: tokio::runtime::coop::with_budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<<mz_adapter::coord::Coordinator>::serve::{closure#0}>::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/coop.rs:107:5
  24: tokio::runtime::coop::budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<<mz_adapter::coord::Coordinator>::serve::{closure#0}>::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/coop.rs:73:5
  25: <tokio::runtime::park::CachedParkThread>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/park.rs:283:31
  26: <tokio::runtime::context::BlockingRegionGuard>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/context.rs:315:13
  27: <tokio::runtime::handle::Handle>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/handle.rs:264:9
  28: mz_adapter::coord::serve::{closure#0}::{closure#1}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b9703a5fb950feb9-1/materialize/tests/src/adapter/src/coord.rs:1549:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

To reproduce, run the attached SQL script in 3 concurrent connections (I have also seen it reproduce with 2). Apologies for the crude test case, I will need to add the DDL fuzzing tools that I use to mzcompose.

```
cat gh18994-item1.sql |  psql -h 127.0.0.1 -p 16875 -U materialize &
cat gh18994-item1.sql |  psql -h 127.0.0.1 -p 16875 -U materialize &
cat gh18994-item1.sql |  psql -h 127.0.0.1 -p 16875 -U materialize &
```
[gh18994-item1.sql.zip](https://github.com/MaterializeInc/materialize/files/11372871/gh18994-item1.sql.zip)

Once this panic is moved out of the way, I will be able to continue testing.

teskje commented on src/compute-client/src/controller.rs:
```diff
@@ -216,6 +216,17 @@ impl<T> ComputeController<T> {
+            instance_id: id,
+            collections: BTreeSet::from_iter(instance.collections_iter().map(|(id, _state)| *id)),
+        })
+    }
```
Is there a reason why this has to live in the compute controller? I'd like to keep the compute controller API as small as possible. In this case it seems like client code already has access to all the available information, since we just collect the IDs of existing collections. So this method and the `ComputeInstanceSnapshot` type could live in adapter, no?

mjibson replied:
Moved as much as I could into adapter. Had to add a method on instance ref.

philip-stoev reviewed (changes requested):
Please ping me when the panic is moved out of the way.

mjibson commented:
I believe the panic is out of the way and this is ready for a retest.

philip-stoev commented:
ok adding it to my todo

ParkMyCar commented on src/adapter/src/coord/dataflows.rs:
```diff
@@ -49,14 +51,43 @@ use crate::session::{Session, SERVER_MAJOR_VERSION, SERVER_MINOR_VERSION};
+/// A reference-less snapshot of a compute instance.
+#[derive(Debug, Clone)]
+pub struct ComputeInstanceSnapshot {
+    instance_id: ComputeInstanceId,
```
naively I'm worried we might run into issues similar to the `ConnectionId` one. i.e. the compute instance doesn't actually exist anymore, but some part of code believes it does since it has a `ComputeInstanceId`. Is this possible? Or do we already assume a compute instance might not exist, even if we have a `ComputeInstanceId`?

mjibson replied:
In the coordinator thread, an instance id won't go away surprisingly since that can also only happen in that thread. Outside of the coord thread, everything has to re-validate when it returns, which happens at the top of each peek stage, which means it'll recheck the instance: https://github.com/MaterializeInc/materialize/blob/cd391bcd33daf9c8651325cc4a28a5cc6ea8f117/src/adapter/src/coord/sequencer/inner.rs#L1992

philip-stoev reviewed (approved):
Unfortunately due to:

- https://github.com/MaterializeInc/materialize/issues/19743
which is also present in main, I am unable to stress-test this as much as I wanted.

So I am hitting Approve here, and the safety of this code when it comes to concurrent DDL will need to be retested at a later time.

mjibson commented:
Thank you. I'll wait on this PR until that issue is fixed since they might overlap.

teskje commented on src/compute-client/src/controller.rs:
```diff
@@ -633,6 +633,11 @@ impl<T> ComputeInstanceRef<'_, T> {
         self.instance.collection(id)
     }
+
+    /// Return a read-only handle to the indicated collection.
```
```suggestion
    /// Return an iterator over the installed collections.
```

mjibson commented:
Revived! Incorporated the previous feedback. Had to add some stuff to accommodate the new optimizer stats. Refactored the DataflowBuilder to also be an index oracle which seems to simplify various functions. It's been enough weeks that I'd appreciate another round of reviews.

Is it also possible to run this against the concurrent workload test?

jkosh44 commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -2034,9 +2034,68 @@ impl Coordinator {
+            .expect("compute instance does not exist");
+        let internal_cmd_tx = self.internal_cmd_tx.clone();
+
+        // TODO: Is there a way to avoid making two dataflow_builders (the second is in
+        // optimize_peek)?
+        let id_bundle = self
+            .dataflow_builder(stage.cluster_id)
+            .sufficient_collections(&stage.source_ids);
```
We never update the `validity` struct with these IDs. Do we need to worry about these becoming invalid while optimize is off thread?

mjibson replied:
Nice catch! Extended the dependent ids from this which should catch it. I'm not sure how `sources.depends_on()` and `sufficient_collections()` differ, but this seems safer and is cheap.

jkosh44 replied:
I believe that `sufficient_collections()` includes indexes while `source.depends_on()` does not.

aalexandrov commented:
> Is it also possible to run this against the concurrent workload test?

@mjibson you can manually start a Nightly BuildKite run against your PR branch ([example](https://buildkite.com/materialize/nightlies/builds?branch=aalexandrov%3Aissue_20557)). In the first stage you can select which Nightlites tests you want to run (presumably only "parallel workload"?).

mjibson commented:
Thanks! Running the parallel workload tests at 58cbbb47c847a4bb92f113508ceb1be13e5e4876.

mjibson commented:
The parellel workload tests passed, going to merge tomorrow unless objections.
